pilot/pkg/serviceregistry/kube/controller/ambient/services.go

```diff
@@ -0,0 +1,130 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// nolint: gocritic
+package ambient
+
+import (
+	v1 "k8s.io/api/core/v1"
+
+	networkingclient "istio.io/client-go/pkg/apis/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pkg/config/schema/kind"
+	"istio.io/istio/pkg/kube/krt"
+	"istio.io/istio/pkg/slices"
+	"istio.io/istio/pkg/workloadapi"
+)
+
+func (a *index) ServicesCollection(
+	Services krt.Collection[*v1.Service],
+	ServiceEntries krt.Collection[*networkingclient.ServiceEntry],
+) krt.Collection[model.ServiceInfo] {
+	ServicesInfo := krt.NewCollection(Services, func(ctx krt.HandlerContext, s *v1.Service) *model.ServiceInfo {
+		portNames := map[int32]model.ServicePortName{}
+		for _, p := range s.Spec.Ports {
+			portNames[p.Port] = model.ServicePortName{
+				PortName:       p.Name,
+				TargetPortName: p.TargetPort.StrVal,
+			}
+		}
+		a.networkUpdateTrigger.MarkDependant(ctx) // Mark we depend on out of band a.Network
```

Comment:
Is it possibly better to pass `ctx` around so the `MarkDependant` can be closer to the place the dependency actually lives?